### PR TITLE
Dblatcher safari fontface add fix

### PR DESCRIPTION
**Pull request by dblatcher**
## What does this change?

 - Addresses the error raised in: [Catch and ignore FontFaceSet InvalidModificationError](https://github.com/guardian/dotcom-rendering/pull/7322)
 - fixes the loading of fonts for the `SecureSignupIframe.importable` component in Safari by cloning the whole`<style>` element containing the font into the iframe if the calls to `FontFaceSet.add` fail,
 - Uses the same clone element method on browsers that don't support `FontFaceSet.add`, rather than falling back to system fonts.

### URL for local testing:
http://localhost:3030/Article/https://www.theguardian.com/games/2022/jul/12/pushing-buttons-is-it-game-over-for-gaming-pandemic-boom

In Safari on main, the error is caught, but the custom fonts do not load in the iframe. On this branch the error is still caught, but the custom fonts are loaded.

## Why?

The Safari behaviour is to throw an `InvalidModificationError` when `FontFaceSet.add` is used to programmatically append a font declaration from the parent document to the iframe, this is specific to Safari as identified in https://github.com/guardian/dotcom-rendering/pull/7322. This meant the guardian fonts were not being added to the iframe in Safari.

Based on the [spec for the method](https://w3c.github.io/csswg-drafts/css-font-loading/#dom-fontfaceset-add), this appears to be a Safari bug. The `InvalidModificationError` indicated that the font being added is "css-connected", IE already declared in the document using a css `@font-face` rule. There are no  `@font-face` rules in the iframe's content document, so this error should not be occurring. It seems that Safari's implementation is checking for the `@font-face` rules on the parent document, rather than the iframe's document. 

The `FontFaceSet.has` and `FontFaceSet.check` methods do not provide the 'false positives' in safari that would predict the `FontFaceSet.add` fails, so using `catch` seems to be the only solution.

Cloning the `<style>` does work for Safari, but the `FontFaceSet.add` is preferable where available as:
 - it allows only the required fonts (in this case, only GuardianTextSans) to be added to the iframe, improving performance
 - It does not rely on the css selector ("style.webfont") to identify the right `<style>` element.

## Screenshots (safari)

The "enter you email address" label is within the iframe. Before, it is in Arial (straight lower case "l" in "email") and After has GuardianTextSans (with curly lowercase "l")

| Before       | After      |
|-------------|------------|
| Screenshot 2023-04-07 at 09 25 19 | Screenshot 2023-04-07 at 09 24 31|

[before]: https://example.com/before.png
[after]: https://example.com/after.png

### CODE REVIEW

File: dotcom-rendering/src/web/lib/addFontsToIframe.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+		return;
+	}
+
+	// get all the fontFaces on the parent matching the list of font names
+	const requiredFonts: FontFace[] = [];
+	parentDocument.fonts.forEach((fontFace) => {
+		if (requiredFontNames.includes(fontFace.family)) {
+			requiredFonts.push(fontFace);
+		}
+	});
```
Comment (AshCorr) [inline dotcom-rendering/src/web/lib/addFontsToIframe.ts:49]:
```suggestion
	const requiredFonts: FontFace[] = parentDocument.fonts.filter((fontFace) => requiredFontNames.includes(fontFace.family)));
```

Nice spot for a filter maybe?

File: dotcom-rendering/src/web/lib/addFontsToIframe.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		} catch (error) {
+			// Safari throws an InvalidModificationError on the add(font) method
+			// https://developer.mozilla.org/en-US/docs/Web/API/FontFaceSet/add#exceptions
+
+			// This may be a Safari bug - the error indicates that the FontFace is already
+			// defining in the document with a css @font-face declaration. The parent window
+			// has @font-face declarations, but the iframe does not.
+
+			usingAddFailed = true;
+			cloneFontFaceElement(parentDocument, iframe);
```
Comment (AshCorr) [inline dotcom-rendering/src/web/lib/addFontsToIframe.ts:68]: After the first exception is thrown we set `usingAddFailed` to `true`, would this not cause any future iterations of this loop to early return on https://github.com/guardian/dotcom-rendering/pull/7533/files#diff-b9cd1532caa176968f4ffc39f3256e54078df602cdb56962580dc82a148049e9R56
thus the font does not get added to the iframe by using `.add` or by cloning (as no exception would be thrown) ?

TLDR no more fonts get added/cloned after one throws an exception?

**Review by AshCorr — APPROVED** [review]
Looks good!

Just a question on the fallback behaviour, I'm not quite sure I understand it properly!

File: dotcom-rendering/src/web/lib/addFontsToIframe.ts (hunk 2 above)
Comment (dblatcher) [reply dotcom-rendering/src/web/lib/addFontsToIframe.ts:68]: It will cause future iterations to early return, but this is intended. `cloneFontFaceElement` copies the whole of the `<style>` element from the parentDocument into the iframe - so if adding one font throws an exception, the whole set of fonts gets copied instead (which we only want to do once).

Or, to put it another way - if one call to `iframeFontFaceSet.add` fails, we assume rest will fail too, so just copy the whole `<style>` element instead.

Looking back at this code, I think it could benefit from some more comments and/or refactoring. It's not very clear.

Comment (AshCorr) [reply dotcom-rendering/src/web/lib/addFontsToIframe.ts:68]: Aha! Gotcha, that makes sense now, thanks!
